crates/component/src/wrappers/http.rs

```diff
@@ -0,0 +1,570 @@
+        stream: OutputStream,
+    ) -> std::io::Result<()> {
+        loop {
+            match stream.blocking_splice(&self, u64::MAX) {
+                Ok(0) | Err(StreamError::Closed) => break,
+                Ok(_) => continue,
+                Err(StreamError::LastOperationFailed(err)) => {
+                    return Err(std::io::Error::new(std::io::ErrorKind::Other, err));
+                }
+            }
+        }
+        drop(stream);
+        wasi::http::types::OutgoingBody::finish(body, None)
+            .map_err(|err| std::io::Error::new(std::io::ErrorKind::Other, err))
+    }
+}
+
+impl OutgoingBody for wasi::http::types::IncomingBody {
+    fn write(
+        self,
+        body: wasi::http::types::OutgoingBody,
+        stream: OutputStream,
+    ) -> std::io::Result<()> {
+        let input = self
+            .stream()
+            .map_err(|()| std::io::Error::from(std::io::ErrorKind::Other))?;
+        loop {
+            match stream.blocking_splice(&input, u64::MAX) {
+                Ok(0) | Err(StreamError::Closed) => break,
+                Ok(_) => continue,
+                Err(StreamError::LastOperationFailed(err)) => {
+                    return Err(std::io::Error::new(std::io::ErrorKind::Other, err));
+                }
+            }
+        }
+        drop(stream);
+        let _trailers = wasi::http::types::IncomingBody::finish(self);
+        // NOTE: getting trailers crashes Wasmtime 25, so avoid doing so
+        //let trailers = if let Some(trailers) = trailers.get() {
+        //    trailers
+        //} else {
+        //    trailers.subscribe().block();
+        //    trailers
+        //        .get()
+        //        .ok_or_else(|| std::io::Error::from(std::io::ErrorKind::Other))?
+        //};
+        //let trailers = trailers.map_err(|()| std::io::Error::from(std::io::ErrorKind::Other))?;
+        //let trailers =
+        //    trailers.map_err(|err| std::io::Error::new(std::io::ErrorKind::Other, err))?;
+        wasi::http::types::OutgoingBody::finish(body, None)
+            .map_err(|err| std::io::Error::new(std::io::ErrorKind::Other, err))
+    }
+}
+
+impl OutgoingBody for IncomingBody {
+    fn write(
+        self,
+        body: wasi::http::types::OutgoingBody,
+        stream: OutputStream,
+    ) -> std::io::Result<()> {
+        loop {
+            match stream.blocking_splice(&self.stream, u64::MAX) {
+                Ok(0) | Err(StreamError::Closed) => break,
+                Ok(_) => continue,
+                Err(StreamError::LastOperationFailed(err)) => {
+                    return Err(std::io::Error::new(std::io::ErrorKind::Other, err));
+                }
+            }
+        }
+        let trailers = self
+            .into_trailers_wasi()
+            .map_err(|err| std::io::Error::new(std::io::ErrorKind::Other, err))?;
+        wasi::http::types::OutgoingBody::finish(body, trailers)
+            .map_err(|err| std::io::Error::new(std::io::ErrorKind::Other, err))
```

Comment:
```suggestion
        }
        let trailers = self
            .into_trailers_wasi()
            .map_err(|err| std::io::Error::new(std::io::ErrorKind::Other, err))?;
        drop(stream);
        wasi::http::types::OutgoingBody::finish(body, trailers)
            .map_err(|err| std::io::Error::new(std::io::ErrorKind::Other, err))
```

Make sure to drop the stream before finishing body here (other `write()` impls all do this)